src/wallet/spend.cpp

```diff
@@ -1392,4 +1392,384 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
+    if (script.IsWitnessProgram(witnessversion, witnessprogram)) {
+        txSize += (unsigned int)(numTxIn * 67.75f + 0.5f);
+    } else {
+        txSize += numTxIn * 148;
+    }
+    return txSize;
+}
+
+float CalculateDeniabilizationProbability(unsigned int deniabilization_cycles)
+{
+    // 100%, 50%, 25%, 13%, 6%, 3%, 2%, 1%
+    return powf(0.5f, deniabilization_cycles);
+}
+
+bool IsDeniabilizationWorthwhile(CAmount total_value, CAmount fee_estimate)
+{
+    constexpr CAmount value_to_fee_ratio = 10;
+    return total_value > fee_estimate * value_to_fee_ratio;
+}
+
+CCoinControl SetupDeniabilizationCoinControl(unsigned int confirm_target)
+{
+    CCoinControl coin_control;
+    coin_control.m_avoid_address_reuse = true;
+    coin_control.m_avoid_partial_spends = true;
+    coin_control.m_allow_other_inputs = false;
+    coin_control.m_signal_bip125_rbf = true;
+    coin_control.m_confirm_target = confirm_target;
+    // we'll automatically bump the fee if economical ends up not confirming by the next deniabilization cycle
+    coin_control.m_fee_mode = FeeEstimateMode::ECONOMICAL;
+    return coin_control;
+}
+
+CFeeRate CalculateDeniabilizationFeeRate(const CWallet& wallet, unsigned int confirm_target)
+{
+    CCoinControl coin_control = SetupDeniabilizationCoinControl(confirm_target);
+
+    CFeeRate requiredFeeRate = GetRequiredFeeRate(wallet);
+    FeeCalculation fee_calc;
+    CFeeRate minFeeRate = GetMinimumFeeRate(wallet, coin_control, &fee_calc);
+    if (fee_calc.reason == FeeReason::FALLBACK || requiredFeeRate > minFeeRate)
+        return requiredFeeRate;
+    return minFeeRate;
+}
+
+static CAmount CalculateDeniabilizationTxFee(const CScript& shared_script, CAmount total_value, unsigned int num_utxos, const CFeeRate& fee_rate)
+{
+    Assert(num_utxos > 0);
+    unsigned int deniabilization_tx_size = CalculateDeniabilizationTxSize(shared_script, total_value, num_utxos);
+    return fee_rate.GetFee(deniabilization_tx_size);
+}
+
+CAmount CalculateDeniabilizationFeeEstimate(const CScript& shared_script, CAmount total_value, unsigned int num_utxos, unsigned int deniabilization_cycles, const CFeeRate& fee_rate)
+{
+    float deniabilizationProbability = CalculateDeniabilizationProbability(deniabilization_cycles);
+    // convert to integer percent to truncate and check for zero probability
+    unsigned int deniabilizationProbabilityPercent = deniabilizationProbability * 100;
+    if (deniabilizationProbabilityPercent == 0) {
+        return 0;
+    }
+
+    // this cycle will use all the UTXOs, while following cycles will have just one UTXO
+    CAmount deniabilizationFee = CalculateDeniabilizationTxFee(shared_script, total_value, num_utxos, fee_rate);
+
+    // calculate the fees from future deniabilization cycles
+    CAmount futureDeniabilizationFee = CalculateDeniabilizationFeeEstimate(shared_script, total_value / NUM_DENIABILIZATION_OUTPUTS, 1, deniabilization_cycles + 1, fee_rate) * 2;
```

Comment:
why compute fee estimate and then multiply by a hardwired 2?

Reply:
Oh, that 2 probably needs to be NUM_DENIABILIZATION_OUTPUTS. Basically we calculate the future cycle cost for each of the chunks we split into. Each chunk should cost the same if the number of splits is the same, so we can calculate one chunk and multiply by the number of splits.
Thanks for spotting that. I'll fix it.

Reply:
happpy to help